### PR TITLE
osd: detect VDO and advertise meaningful utilization numbers

### PR DESCRIPTION
**Pull request by liewegas**
VDO is a thinly provisioned device, which means our real OSD utilization depends on the size and utilization of the VDO volume's physical device as well as the virtual device layered over it.  Free space isthe min of the two.

TODO
- [x] allow vdo code to look "through" LV devices to underlying VDO
- [ ] test with ceph-volume and ensure it detects VDO properly

### CODE REVIEW

**Comment by liewegas** [comment]
FWIW I got VDO up and running quickly and easily on my fedora 27 box (thanks to @rootfs) with
```
sudo dnf copr enable rhawalsh/dm-vdo
sudo dnf install vdo kmod-kvdo 
```
I also had to update my kernel version to match the module in kmod-kvdo.

**Comment by liewegas** [comment]
@jdurgin I have this working locally (with a manually configured VDO device and LV on top of it).

Once ceph-volume support is there we'll need to do some performance and failure testing before blessing it...

**Comment by alfredodeza** [comment]
hrmn, we can't have support in ceph-volume until ceph-ansible has support for vdo really, since ceph-volume doesn't do any kind of config management here, merely accepts LVs or partitions. We also do all our verification with ceph-ansible, so support has to exist there for us to validate

**Comment by liewegas** [comment]
Oh, right.  In that case, is there any ceph-volume change needed at all?

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -6404,31 +6404,48 @@ int BlueStore::get_devices(set<string> *ls)
 int BlueStore::statfs(struct store_statfs_t *buf)
 {
   buf->reset();
-  buf->total = bdev->get_size();
-  buf->available = alloc->get_free();
 
+  uint64_t free = alloc->get_free();
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc]: aliasing free() is a bit confusing

File: src/common/blkdev.cc [hunk 2]
```diff
@@ -297,6 +300,76 @@ void get_dm_parents(const std::string& dev, std::set<std::string> *ls)
...
+{
+  int64_t ret = 0;
+  int fd = ::openat(vdo_fd, property, O_RDONLY);
+  if (fd < 0) {
+    return 0;
+  }
+  char buf[1024];
+  int r = ::read(fd, buf, sizeof(buf));
+  if (r > 0) {
+    buf[r] = 0;
```
Comment (jdurgin) [inline src/common/blkdev.cc]: off by one - past the end of the buffer if r == sizeof(buf)

File: src/os/filestore/FileStore.cc [hunk 3]
```diff
@@ -735,8 +735,17 @@ int FileStore::statfs(struct store_statfs_t *buf0)
     assert(r != -ENOENT);
     return r;
   }
-  buf0->total = buf.f_blocks * buf.f_bsize;
-  buf0->available = buf.f_bavail * buf.f_bsize;
+
+  uint64_t free = buf.f_bavail * buf.f_bsize;
```
Comment (jdurgin) [inline src/os/filestore/FileStore.cc]: aliasing free()

**Comment by jdurgin** [comment]
looks good in general, just a few minor comments

**Comment by serjponomarev** [comment]
@liewegas 
There is a problem, Ceph does not correctly calculate the RAW space for OSD based on VDO.
The problem apparently arose after integrating VDO into LVM.

https://tracker.ceph.com/issues/63674

**Pull request merged**